src/r_voxel.c

```diff
@@ -0,0 +1,962 @@
+			best_dist = dist;
+		}
+	}
+
+	return best;
+}
+
+
+static void VX_CreateRemapTable (byte * p, byte * table)
+{
+	byte * pal = W_CacheLumpName ("PLAYPAL", PU_CACHE);
+
+	int c;
+	for (c = 0 ; c < 256 ; c++)
+	{
+		int r = (int)*p++ << 2;
+		int g = (int)*p++ << 2;
+		int b = (int)*p++ << 2;
+
+		table[c] = VX_PaletteIndex (pal, r, g, b);
+	}
+}
+
+
+static void VX_RemapSlabColors (struct Voxel * v, int x, int y, byte * table)
+{
+	int A = v->offsets[y     * v->x_size + x];
+	int B = v->offsets[(y+1) * v->x_size + x];
+
+	if (! (A < B))
+		return;
+
+	byte * slab = &v->data[A];
+	byte * end  = &v->data[B];
+
+	while (slab < end)
+	{
+		byte top  = *slab++;
+		byte len  = *slab++;
+		byte face = *slab++;
+		byte i;
+
+		(void) top;
+		(void) face;
+
+		for (i = 0 ; i < len ; i++, slab++)
+		{
+			*slab = table[*slab];
+		}
+	}
+}
+
+
+static struct Voxel * VX_Decode (byte * p, int length)
+{
+	// too short?
+	if (length < 40 + 768)
+		return NULL;
+
+	byte * orig_p = p;
+
+	struct Voxel * v = Z_Malloc (sizeof(struct Voxel), PU_STATIC, NULL);
+
+	// skip num_bytes
+	p += 4;
+
+	v->x_size = (int)p[0];  p += 4;
+	v->y_size = (int)p[0];  p += 4;
+	v->z_size = (int)p[0];  p += 4;
+
+	if (v->x_size == 0 || v->y_size == 0 || v->z_size == 0)
+		return NULL;
+
+	v->x_pivot = (p[0] << 8) | (p[1] << 16);  p += 4;
+	v->y_pivot = (p[0] << 8) | (p[1] << 16);  p += 4;
+	v->z_pivot = (p[0] << 8) | (p[1] << 16);  p += 4;
+
+	// decode the offset values...
+	int xoffsets[260];
+
+	int x;
+	for (x = 0 ; x <= v->x_size ; x++)
+	{
+		xoffsets[x] = (int)p[0] | (p[1] << 8) | (p[2] << 16);
+		p += 4;
+	}
+
+	int num_offsets = v->x_size * (v->y_size + 1);
+
+	int min_offset = (1 << 30);
+	int max_offset = 0;
+
+	v->offsets = Z_Malloc (sizeof(int) * num_offsets, PU_STATIC, NULL);
+
+	for (x = 0 ; x < v->x_size ; x++)
+	{
+		int y;
+		for (y = 0 ; y <= v->y_size ; y++)
+		{
+			int offset = (int)p[0] | (p[1] << 8);
+			p += 2;
+
+			offset = offset + xoffsets[x];
+
+			v->offsets[y * v->x_size + x] = offset;
+
+			if (offset < min_offset) min_offset = offset;
+			if (offset > max_offset) max_offset = offset;
+		}
+	}
+
+	int data_size = max_offset - min_offset;
+	if (data_size <= 0)
+		return NULL;
+
+	for (x = 0 ; x < num_offsets ; x++)
+		v->offsets[x] -= min_offset;
+
+	// copy the slab data
+	p = orig_p + (7 * 4) + min_offset;
+
+	v->data = Z_Malloc (data_size, PU_STATIC, NULL);
+
+	memcpy (v->data, p, data_size);
+
+	// handle palette: create a mapping table
+	byte remap_table[256];
+
+	VX_CreateRemapTable (orig_p + (length - 768), remap_table);
+
+	// remap colors...
+	for (x = 0 ; x < v->x_size ; x++)
+	{
+		int y;
+		for (y = 0 ; y < v->y_size ; y++)
+		{
+			VX_RemapSlabColors (v, x, y, remap_table);
+		}
+	}
+
+	return v;
+}
+
+
+static boolean VX_Load (int spr, int frame)
+{
+	char frame_ch = 'A' + frame;
+
+	if (frame_ch == '\\')
+		frame_ch = '^';
+
+	char filename[256];
+
+	M_snprintf (filename, sizeof(filename), "voxels/%s%c.kvx", sprnames[spr], frame_ch);
```

Comment:
How are we going to handle the data loading? Autoload directory? Command line parameter?

Reply:
If you don't mind me chiming in, I was thinking about a `VX_` namespace in WADs, to make this more viable for modders.

Reply:
I think we should add support for the `.kvx` extension and load it from `autoload` or `.zip` files. Maybe we should also support the `voxels` directory.

Reply:
> If you don't mind me chiming in, I was thinking about a `VX_` namespace in WADs, to make this more viable for modders.

That's a good idea too. I don't think anyone supports voxels in WADs, right?

Reply:
> That's a good idea too. I don't think anyone supports voxels in WADs, right?

I mean, I don't know if other ports would ignore stuff between markers they don't recognize. Optimally, they should load e.g. a `BON1A0` between `S_` (or `SS_`) markers but not a `BON1A0` between something like these `VX_` markers I'm proposing.

On that note, with voxels we don't have to worry about rotations, so I think another possibility would be to include voxels in WADs by giving them a prefix: `VOX[sprite]`, e.g. `VOXBON1A`. That shouldn't conflict with normal sprites and probably wouldn't need markers.

Reply:
I believe ports like pr+ and dsda-doom _would_ load stuff between markers that are unknown.

Reply:
Sounds like that'd require a cross-port effort, which I'd personally rather avoid. So, using the exact names of the sprites we're providing voxels for is a no-go.

Perhaps we could combine both of the approaches I proposed: use a `VX_` namespace to tell our voxel-supporting port that the lumps in that namespace are indeed voxels, and also use a prefix or suffix to not confuse non-voxel-supporting ports into trying to load voxels as sprites.

Maybe the prefix/suffix part is unnecessary; I'm also not sure if ports try to load sprites outside of the `S_` namespace.

Reply:
The *.kvm files have no headers, so identifying them is rather tedious: https://github.com/ZDoom/gzdoom/blob/d5e9783324936b50eb4b7b8ed646894a30e226b2/src/common/models/voxels.cpp#L168
So I think load from WADs and *.lmp is not a good idea.

With the last commit we can load *.kvm from *.zip and *.pk3, -file parameter, autoload directory and `voxels` directory. `cheello_voxels_v2_1.pk3` just works.

Reply:
> With the last commit we can load *.kvm from *.zip and *.pk3, -file parameter, autoload directory and `voxels` directory. `cheello_voxels_v2_1.pk3` just works.

This is the best solution, awesome! 👍